### PR TITLE
Make rextendr packages CRAN compatible by default

### PR DESCRIPTION
**Pull request by JosiahParry**
As a followup from #393 and comments from @CGMossa and @Ilia-Kosenkov, this PR makes _adds_ `--print=native-static-libs` to any existing `RUSTFLAGS` ensuring that they are not overwritten.

### CODE REVIEW

**Comment by JosiahParry** [comment]
` Portable Makefiles do not use GNU extensions such as +=, :=,` 

welp! back to the drawing board!

**Comment by Ilia-Kosenkov** [comment]
Can't you do something like `RUSTFLAGS = $(RUSTFLAGS) --whatever`?
UPD: ChatGPT suggests this (did not test it myself)
```
RUSTFLAGS := $(RUSTFLAGS) --print=native-static-libs
export RUSTFLAGS
```

File: inst/templates/Makevars.in [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	RUSTFLAGS="$(RUSTFLAGS) --print=native-static-libs" cargo build @CRAN_FLAGS@ --lib --release --manifest-path=./rust/Cargo.toml --target-dir $(TARGET_DIR)
+
+	# Always clean up CARGOTMP
+	rm -Rf $(CARGOTMP)
+
+ 	if [ "$(NOT_CRAN)" != "false" ]; then \
+        rm -Rf $(VENDOR_DIR); \
+        rm -Rf $(TARGET_DIR); \
+        rm -Rf $(LIBDIR)/build; \
+    fi
```
Comment (JosiahParry) [inline inst/templates/Makevars.in:44]: I think i have this logic backwards....I think we _want_ to remove these directories when `NOT_CRAN="false"`? Either way both `R CMD checks` pass so maybe it isn't needed?

File: tests/testthat/test-cran-compliance.R [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+  package_versions <- vendor_pkgs(path, quiet = TRUE)
+  expect_snapshot(cat_file("src", "rust", "vendor-config.toml"))
+  expect_snapshot(package_versions, transform = mask_any_version)
+  expect_true(file.exists(file.path("src", "rust", "vendor.tar.xz")))
+})
+
+
+test_that("rextendr passes NOT_CRAN=false checks", {
+    skip_if_not_installed("usethis")
```
Comment (JosiahParry) [inline tests/testthat/test-cran-compliance.R:17]: This test ensures that `R CMD check` is passed when `NOT_CRAN=false`

File: tests/testthat/test-use_extendr.R [hunk 3]
```diff
@@ -184,3 +184,21 @@ test_that("`use_extendr()` works correctly when path is specified explicitly", {
   use_extendr(path = "testpkg")
   succeed()
 })
+
+
+test_that("`use_extendr()` passes R CMD check", {
+  skip_if_not_installed("usethis")
+  skip_if_not_installed("rcmdcheck")
```
Comment (JosiahParry) [inline tests/testthat/test-use_extendr.R:191]: This test ensures that `R CMD check` passes when not on CRAN.

**Review by JosiahParry — COMMENTED** [review]
@Ilia-Kosenkov @CGMossa 

I've taken this as an opportunity to refactor the way in which Makevars are done.

## Summary 

This PR now removes `use_cran_defaults()` and instead, uses a single `Makevars` and `Makevars.in` to support CRAN environment and regular development environment. This is done in the following way: 

- `Makevars` & `Makevars.win` are now `Makevars.in` and `Makevars.win.in`
- The `configure` script generates the `Makevars` and `Makevars.win` files substituting the `@CRAN_FLAGS@` this approach is taken from `data.table` see https://github.com/Rdatatable/data.table/blob/master/src/Makevars.in. It also deletes the `src/Makevars` and `src/Makevars.win` before generating the respective files
- Now `src/Makevars{.win}` are in the `.Rbuildignore` and `.gitignore` when using `use_extendr()` ensuring that the file is always rebuilt
- When `NOT_CRAN=false` the `vendor.tar.xz` is unbundled and the flags `-j 2 --offline` are set

To ensure all of this is done correctly, two new tests are added which run `rcmdcheck::rcmdcheck()` and `rcmdcheck::rcmdcheck(env = c("NOT_CRAN"="false"))` to check both scenarios.

The function `use_cran_defaults()` has been removed. 

## Next steps

If you both approve of this direction, I would like to make the following changes 

### Within this PR
 
- [ ] remove `vignettes/articles/cran-compliance.qmd` as an alternative i am happy to write a standalone help doc e.g. `?rextendr::cran`
- [x] remove mentions of `use_cran_defaults()` from the NEWS.md as the function will have been introduced and removed before a new CRAN release

### Follow up PRs

- [ ] write a function to generate an `inst/AUTHORS` file as was done by @DavZim in {rtiktoken} see https://github.com/DavZim/rtiktoken/blob/master/inst/AUTHORS
- [ ] write a page in the user guide on publishing to CRAN

**Comment by Ilia-Kosenkov** [comment]
I still believe checking `NOT_CRAN` for anything other than `true` is invalid. `NOT_CRAN` is an optional env var that you can set to, usually, `true` to signify that you are not on CRAN. In all other scenarios, you treat is as if you are on CRAN. What does `NOT_CRAN=false` mean? When do you expect to set this?

**Comment by JosiahParry** [comment]
Thanks @Ilia-Kosenkov, that's fine. That can be addressed. But are you okay with this plan (assuming we change the NOT_CRAN to check based on a true value)? Having one ~~ring~~ Makevars to rule them all?

**Comment by DavZim** [comment]
What happens when NOT_CRAN is true but I don't have vendor/ only vendor.tar.xz? Ie I installed the package from GitHub or from local.

Is the archive unpacked correctly?
Btw, once this is done, im happy to upgrade the other blog accordingly.

**Comment by JosiahParry** [comment]
@DavZim the Makevars is only ran when building the package from source. This template will unpack the archive if `NOT_CRAN` is unset *and* the `vendor.tar.xz` file exists. If `NOT_CRAN=true` the case when using devtools etc then the archive is not unpacked.

**Review by CGMossa — APPROVED** [review]

File: inst/templates/Makevars.in [hunk 4]
```diff
@@ -0,0 +1,44 @@
+TARGET_DIR = ./rust/target
+LIBDIR = $(TARGET_DIR)/release
+STATLIB = $(LIBDIR)/lib{{{lib_name}}}.a
```
Comment (CGMossa) [inline inst/templates/Makevars.in:3]: Unrelated to things, but I think this could be named `RUSTLIB` without any issue..

Comment (JosiahParry) [reply inst/templates/Makevars.in:3]: Do you think doing that could help the possibility of including Rcpp + extendr in the same package?

**Comment by JosiahParry** [comment]
I've moved this as a draft so I can adjust the `configure` scripts to set `CRAN_FLAGS` when `NOT_CRAN` is unset **and** a `vendor.tar.xz` folder is present.

File: inst/templates/Makevars.in (hunk 4 above)
Comment (CGMossa) [reply inst/templates/Makevars.in:3]: Let's have a meeting about this at some point. I think I know what needs to be investigated to do this.

**Pull request merged**